src/common/lru.c

```diff
@@ -289,15 +292,17 @@ daos_lru_ref_evict_wait(struct daos_lru_cache *lcache, struct daos_llink *llink)
 	if (!llink->ll_evicted)
 		daos_lru_ref_evict(lcache, llink);
 
-	if (lcache->dlc_ops->lop_wait && !daos_lru_is_last_user(llink)) {
-		/* Wait until I'm the last one.
-		 * XXX: the implementation can only support one waiter for now, if there
-		 * is a secondary ULT calls this function on the same item, it will hit
-		 * the assertion.
-		 */
-		D_ASSERT(!llink->ll_wait_evict);
-		llink->ll_wait_evict = 1;
+	/* - hash table has +1 ll_ref, myself also has +1 ll_ref (reason of 2).
+	 * - each eviction caller has +1 ll_ref, also has +1 ll_wait_evict.
+	 * - each active user has +1 ll_ref.
+	 *
+	 * if (ll_ref - ll_wait_evict) == 2, it means there is no active user on
+	 * container, I don't need to wait anymore.
+	 */
```

Comment:
That reminds me why we had a "ll_evicting" state before.

Current implementation will make the cont_child as "evicted" on the first daos_lru_ref_evict_wait() caller, so that the second one who tries to destroy container will fail to find the cont_child and continue the destroy. (see cont_child_destroy_one() -> cont_child_lookup()).

I'm wondering why do we want to support concurrent container destroy? Even if two concurrent both hold cont_child and wait in daos_lru_ref_evict_wait(), one of it will eventually fail on later container destroy operations, isn't it?

Shouldn't we make it simpler to fail the second container destroy on daos_lru_ref_hold() (by checking the "evicting" state)? (daos_lru_ref_evict_wait() needs be slightly modified to ignore or fail the second caller).

Reply:
I thought about this, but I didn't  choose this way because the second destroy() will return success even the first one is still in progress, which is confusing because the actual destroy is not done yet. We are not supposed to return success to either of callers.

Reply:
The second destroy should fail on cont_child_lookup() & daos_lru_ref_evict_wait(), since it'll find the item is in 'evicting' state.

Your current approach will allow the second destroy moving forward to destroy the container even when the container is still in use. (since the first destroy blindly set the item as 'evicted' despite of the refcount)